pkg/connector/roles.go

```diff
@@ -127,44 +149,125 @@ func (o *roleBuilder) Grants(
 	annotations.Annotations,
 	error,
 ) {
-	var outputAnnotations annotations.Annotations
-	page, limit, _, err := client.ParsePaginationToken(token)
+	var bag pagination.Bag
+
+	err := bag.Unmarshal(token.Token)
 	if err != nil {
 		return nil, "", nil, err
 	}
 
-	users, total, ratelimitData, err := o.client.GetRoleUsers(
-		ctx,
-		resource.Id.Resource,
-		limit,
-		page,
-	)
-	outputAnnotations.WithRateLimiting(ratelimitData)
-	if err != nil {
-		return nil, "", outputAnnotations, err
-	}
+	state := bag.Pop()
+	if state == nil {
+		bag.Push(pagination.PageState{
+			Token:          "",
+			ResourceTypeID: userResourceType.Id,
+		})
 
-	if len(users) == 0 {
-		return nil, "", outputAnnotations, nil
+		if o.syncPermissions {
+			bag.Push(pagination.PageState{
+				Token:          "",
+				ResourceTypeID: scopeResourceType.Id,
+			})
+		}
+
+		nextToken, err := bag.Marshal()
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		return nil, nextToken, nil, nil
 	}
 
-	var grants []*v2.Grant
-	for _, user := range users {
-		principalId, err := resourceSdk.NewResourceID(userResourceType, user.UserId)
+	switch state.ResourceTypeID {
+	case userResourceType.Id:
+		var outputAnnotations annotations.Annotations
+		page, limit, _, err := client.ParsePaginationTokenString(state.Token)
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		users, total, ratelimitData, err := o.client.GetRoleUsers(
+			ctx,
+			resource.Id.Resource,
+			limit,
+			page,
+		)
+		outputAnnotations.WithRateLimiting(ratelimitData)
 		if err != nil {
 			return nil, "", outputAnnotations, err
 		}
-		nextGrant := grant.NewGrant(
-			resource,
-			roleEntitlementName,
-			principalId,
+
+		if len(users) == 0 {
+			return nil, "", outputAnnotations, nil
+		}
+
+		var grants []*v2.Grant
+		for _, user := range users {
+			principalId, err := resourceSdk.NewResourceID(userResourceType, user.UserId)
+			if err != nil {
+				return nil, "", outputAnnotations, err
+			}
+			nextGrant := grant.NewGrant(
+				resource,
+				roleEntitlementName,
+				principalId,
+			)
+			grants = append(grants, nextGrant)
+		}
+
+		bag.Push(pagination.PageState{
+			Token:          client.GetNextToken(page, limit, total),
+			ResourceTypeID: userResourceType.Id,
+		})
+
+		nextToken, err := bag.Marshal()
+		if err != nil {
+			return nil, "", nil, err
```

Comment:
nit: I think you can do `bag.Next` for this. It will keep the current resource type

Reply:
In this case I can't use it, since I use pop before.

The api for bag is ambiguous since it can leak so many usages; it does not enforce one usage